Automate docs

forsyth2 opened the pull request:
Automate docs. Resolves #51.

This is technically for `v1.0.0`. However, since none of the changes should affect `zppy` users, it should be fine to merge after the release of `v1.0.0`.

forsyth2 commented:
@tomvothecoder How can I test this locally? Build the conda environment and try `make html`? How can I test the CI/CD is working?

I addressed everything on your task list in #51.

tomvothecoder commented:
> @tomvothecoder How can I test this locally? Build the conda environment and try `make html`? How can I test the CI/CD is working?
> 
> I addressed everything on your task list in #51.

To test docs locally without a version dropdown, build conda env and `make html` in `docs/`.

To test docs locally with a version dropdown, you need to add `sphinx-multiversion` in the `dev.yml` and follow 6. in the [diags docs](https://e3sm-project.github.io/e3sm_diags/_build/html/master/contributing.html#edit-documentation).

To test doc builds in CI/CD, you need to:
1) Have a `gh-pages` branch on your fork
2) Have GH pages configured to point to `gh-pages`
3) Temporarily disable [line 69](https://github.com/E3SM-Project/zppy/pull/83/files#diff-febd9b6ea2b89a1713ed28b8fe8ee1ca70bce86d9ab532f2fcc4ba9920b202cbR69) of  `build-workflow.yml` so that PRs can trigger the `publish-docs` job. 
4) Check CI/CD passes and job is successful
5) Check https://forsyth2.github.io/zppy/ is built
6) Re-enable line 69

After you confirm it works, you won't need to disable and re-enable line 69 moving forward.

tomvothecoder commented:
You're missing a few things to get `sphinx-multiversion`:
1) Add [`docs/source/_templates/versions.html`](https://github.com/tomvothecoder/e3sm_diags/blob/master/docs/source/_templates/versions.html). This Jinja template renders the drop down box.
2) Follow [this](https://holzhaus.github.io/sphinx-multiversion/master/quickstart.html) for adding `conf.py` settings

tomvothecoder commented:
You're missing a few things to get `sphinx-multiversion` working:
1) Add [`docs/source/_templates/versions.html`](https://github.com/tomvothecoder/e3sm_diags/blob/master/docs/source/_templates/versions.html). This Jinja template renders the drop down box.
2) Follow [sphinx-multiversion quickstart](https://holzhaus.github.io/sphinx-multiversion/master/quickstart.html) for `docs/source/conf.py` settings

forsyth2 commented on .github/workflows/build_workflow.yml:
```diff
@@ -64,3 +64,63 @@ jobs:
           python -m unittest tests/test_*.py
+
+  publish-docs:
+    #if: ${{ github.event_name == 'push' }}
```
Undo this change

forsyth2 commented on .github/workflows/release_workflow.yml:
```diff
@@ -54,3 +54,62 @@ jobs:
+          branch: gh-pages
+          directory: gh-pages
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+          force: true
```
Add extra line

forsyth2 commented:
@tomvothecoder Thanks for the guidance. I believe I've addressed all your comments. I created https://github.com/forsyth2/zppy/pull/1. The 3 checks passed on that branch before merging. I then merged but I didn't see any more checks appear. I checked that https://forsyth2.github.io/zppy/_build/html/main/index.html built successfully as well. 

I see the `publish-docs` check fails on this PR. If that's expected and everything else looks fine to you, I can merge.

tomvothecoder commented:
@forsyth2 Awesome, glad to hear things went smoothly. Just a heads up, merging the PR to your fork triggered the [build workflow](https://github.com/forsyth2/zppy/actions/runs/966228530) on `main` which ran `publish-docs`.

You'll be good to go once you uncomment line #69 so that `publish-docs` is skipped for this and all subsequent PRs.
